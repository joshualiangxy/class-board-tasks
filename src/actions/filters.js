export const sortByDeadline = () => ({ type: 'SORT_BY_DEADLINE' });

export const sortByName = () => ({ type: 'SORT_BY_NAME' });

export const setTextFilter = (text = '') => ({ type: 'SET_TEXT_FILTER', text });

export const toggleGrouped = () => ({ type: 'TOGGLE_GROUPED' });
