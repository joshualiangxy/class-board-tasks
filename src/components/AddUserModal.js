import React, { useState } from 'react';
import Modal from 'react-modal';
import firebase from '../firebase/firebase';

// Note: This still works off uid for now. This also does NOT work yet.
const AddUserModal = ({ isOpen, onRequestClose, group }) => {
  const addNewUser = (userEmail, group) => {
    const uid = userEmail; // TODO: Change this when email-uid are linked!

    // Both of these references are DocumentReference
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(group);

    // Gets the data from a user.
    userRef.get().then(doc => {
      const data = doc.data();
      // Write displayName, studentNum, and UID into a new document
      if (typeof data === 'undefined') {
        // Most likely a user that doesn't exist.
        setError('No such user found');
      } else {
        groupRef
          .collection('users')
          .doc(uid) // Create document for this user's UID
          .set({
            studentNum: data.studentNum,
            displayName: data.displayName,
            uid: uid,
            admin: false
          })
          .then(() => {
            setError('User added!');
          })
          .catch(error => {
            setError('Something went wrong.');
          });
      }
      // Defaults to non-admin.
    });
  };

  // TODO: When calling this modal, pass in the group UID as a prop
  const [userEmail, setUserEmail] = useState('');
  const [error, setError] = useState('');
  // const [usersAdded, setUsersAdded] = useState([]); // For listing down people that have been added?

  const onUserEmailChange = e => {
    setUserEmail(e.target.value);
  };

  const onCancel = () => {
    setUserEmail('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedEmail = userEmail.trim();
    if (!submittedEmail) {
      setError('Please enter a group name');
    } else {
      addNewUser(submittedEmail, group);
      // TODO: Catch user not found
      // Does not close after submitting, so that the user can continue to add more users

      // setUserEmail(''); // Also resets the field.
      // setError('Added!');
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="User ID (required)"
          value={userEmail}
          onChange={onUserEmailChange}
          autoFocus
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

export default AddUserModal;
