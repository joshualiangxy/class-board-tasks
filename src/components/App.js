import React from 'react';
import '../styles/App.css';

const App = () => (
  <div>
    <h1>This is my app</h1>
  </div>
);

export default App;

