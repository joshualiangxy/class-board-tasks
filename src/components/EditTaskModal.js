import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import {
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks
} from '../actions/tasks';
import TaskForm from './TaskForm';

export const EditTaskModal = ({
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  task,
  users,
  editGroupTask
}) => {
  const submitTask = updates =>
    gid
      ? startEditGroupTask(task.id, updates, groupName).then(() => {
          editGroupTask(task.id, updates);
          startSetTasks(); // Temporary solution
          onRequestClose();
        })
      : startEditPersonalTask(task.id, updates).then(() => onRequestClose());

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="Edit Task"
      onRequestClose={onRequestClose}
      appElement={document.getElementById('root')}
    >
      <h2>Edit Task</h2>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
        task={task}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startEditPersonalTask: (id, updates) =>
    dispatch(startEditPersonalTask(id, updates)),
  startEditGroupTask: (id, updates, groupName) =>
    dispatch(startEditGroupTask(id, updates, groupName)),
  startSetTasks: () => dispatch(startSetTasks())
});

export default connect(undefined, mapDispatchToProps)(EditTaskModal);
