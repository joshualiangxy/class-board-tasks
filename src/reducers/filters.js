const defaultState = { text: '', sortBy: 'deadline', grouped: true };

const filtersReducer = (state = defaultState, action) => {
  switch (action.type) {
    case 'SORT_BY_NAME':
      return { ...state, sortBy: 'name' };
    case 'SORT_BY_DEADLINE':
      return { ...state, sortBy: 'deadline' };
    case 'SET_TEXT_FILTER':
      return { ...state, text: action.text };
    case 'TOGGLE_GROUPED':
      return { ...state, grouped: !state.grouped };
    default:
      return state;
  }
};

export default filtersReducer;
