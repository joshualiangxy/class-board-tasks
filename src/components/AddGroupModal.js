import React, { useState } from 'react';
import Modal from 'react-modal';
import { startNewGroup } from '../actions/groups';
import { connect } from 'react-redux';

// TODO: Add the ability to add other users into the group
const AddGroupModal = ({ isOpen, onRequestClose, startNewGroup }) => {
  const [groupName, setGroupName] = useState('');
  const [module, setModule] = useState('');
  const [error, setError] = useState('');

  const onGroupNameChange = e => setGroupName(e.target.value);

  const onModuleChange = e => {
    const module = e.target.value.toUpperCase();

    setModule(module);
  };

  const onCancel = () => {
    setError('');
    setGroupName('');
    setModule('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedName = groupName.trim();
    const submittedMod = module.trim();
    if (!submittedName) {
      setError('Please enter a group name');
    } else {
      setError('');
      setGroupName('');
      setModule('');
      startNewGroup(submittedName, submittedMod).then(() => onRequestClose());
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={onCancel}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Group name (required)"
          value={groupName}
          onChange={onGroupNameChange}
          autoFocus
        />
        <input
          type="text"
          placeholder="Module code/Class name (Optional)"
          value={module}
          onChange={onModuleChange}
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startNewGroup: (groupName, module) =>
    dispatch(startNewGroup(groupName, module))
});

export default connect(null, mapDispatchToProps)(AddGroupModal);
