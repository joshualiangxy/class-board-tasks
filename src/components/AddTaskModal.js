import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startAddPersonalTask, startAddGroupTask } from '../actions/tasks';
import TaskForm from './TaskForm';

export const AddTaskModal = ({
  startAddPersonalTask,
  startAddGroupTask,
  removeAdmin,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  addGroupTask,
  users
}) => {
  const submitTask = task =>
    gid
      ? startAddGroupTask(task, groupName)
          .then(isAdmin => {
            if (isAdmin) return addGroupTask(task);
            else return removeAdmin();
          })
          .then(() => onRequestClose())
      : startAddPersonalTask(task).then(() => onRequestClose());

  return (
    <Modal
      className="task-modal"
      isOpen={isOpen}
      contentLabel="Add Task"
      onRequestClose={onRequestClose}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <h1>Add Task</h1>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startAddPersonalTask: task => dispatch(startAddPersonalTask(task)),
  startAddGroupTask: (task, groupName) =>
    dispatch(startAddGroupTask(task, groupName))
});

export default connect(undefined, mapDispatchToProps)(AddTaskModal);
