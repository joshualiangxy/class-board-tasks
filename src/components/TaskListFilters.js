import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, toggleGrouped } from '../actions/filters';

export const TaskListFilters = ({ filters, setTextFilters, toggleGrouped }) => {
  const onTextChange = e => setTextFilters(e.target.value);

  return (
    <div className="content-container">
      <div className="input-group">
        <div className="input-group__item">
          <input
            className="text-input"
            type="text"
            value={filters.text}
            onChange={onTextChange}
            placeholder="Search tasks"
          />
        </div>
        <div className="input-group__item">
          <label className="switch">
            <input
              className="toggle"
              type="checkbox"
              checked={filters.grouped}
              onChange={toggleGrouped}
            />
            <span className="slider"></span>
          </label>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = ({ filters }) => ({ filters });

const mapDispatchToProps = dispatch => ({
  setTextFilters: text => dispatch(setTextFilter(text)),
  toggleGrouped: () => dispatch(toggleGrouped())
});

export default connect(mapStateToProps, mapDispatchToProps)(TaskListFilters);
