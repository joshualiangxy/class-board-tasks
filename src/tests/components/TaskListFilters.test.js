import React from 'react';
import { shallow } from 'enzyme';
import { filtersDeadline, filtersName } from '../fixtures/filters';
import { TaskListFilters } from '../../components/TaskListFilters';

const setTextFilters = jest.fn();
const toggleGrouped = jest.fn();
let wrapper;

beforeEach(() => {
  jest.clearAllMocks();

  wrapper = shallow(
    <TaskListFilters
      filters={filtersDeadline}
      setTextFilters={setTextFilters}
      toggleGrouped={toggleGrouped}
    />
  );
});

describe('render', () => {
  it('should render TaskListFilters', () => expect(wrapper).toMatchSnapshot());

  it('should render TaskListFilters with alt filters', () => {
    wrapper.setProps({ filters: filtersName });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('text filter', () => {
  it('should call setTextFilters on input change', () => {
    const value = 'filter';

    expect(wrapper.find('input').at(0).prop('value')).toBe(
      filtersDeadline.text
    );

    wrapper.find('input').at(0).prop('onChange')({ target: { value } });

    expect(setTextFilters).toHaveBeenCalledTimes(1);
    expect(setTextFilters).toHaveBeenLastCalledWith(value);
  });
});

describe('grouped checkbox', () => {
  it('should toggle grouped state on change', () => {
    expect(wrapper.find('input').at(1).prop('checked')).toBe(
      filtersDeadline.grouped
    );

    wrapper.find('input').at(1).prop('onChange')();

    expect(toggleGrouped).toHaveBeenCalledTimes(1);
  });
});
