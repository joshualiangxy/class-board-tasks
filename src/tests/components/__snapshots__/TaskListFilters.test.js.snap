// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render TaskListFilters 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Search tasks"
        type="text"
        value=""
      />
    </div>
    <div
      className="input-group__item"
    >
      <label
        className="switch"
      >
        <input
          checked={false}
          className="toggle"
          onChange={[MockFunction]}
          type="checkbox"
        />
        <span
          className="slider"
        />
      </label>
    </div>
  </div>
</div>
`;

exports[`render should render TaskListFilters with alt filters 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Search tasks"
        type="text"
        value=""
      />
    </div>
    <div
      className="input-group__item"
    >
      <label
        className="switch"
      >
        <input
          checked={false}
          className="toggle"
          onChange={[MockFunction]}
          type="checkbox"
        />
        <span
          className="slider"
        />
      </label>
    </div>
  </div>
</div>
`;
